OrbitClientModel/CaptureDeserializerTest.cpp

```diff
@@ -0,0 +1,408 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <vector>
+
+#include "OrbitClientModel/CaptureDeserializer.h"
+#include "absl/base/casts.h"
+#include "capture_data.pb.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using orbit_client_protos::CallstackEvent;
+using orbit_client_protos::CallstackInfo;
+using orbit_client_protos::CaptureHeader;
+using orbit_client_protos::CaptureInfo;
+using orbit_client_protos::FunctionInfo;
+using orbit_client_protos::LinuxAddressInfo;
+using orbit_client_protos::TimerInfo;
+using orbit_client_protos::TracepointEventInfo;
+using orbit_grpc_protos::TracepointInfo;
+
+using ::testing::_;
+using ::testing::Assign;
+using ::testing::DoAll;
+using ::testing::HasSubstr;
+using ::testing::IsEmpty;
+using ::testing::Matcher;
+using ::testing::NotNull;
+using ::testing::SaveArg;
+
+using ::testing::InvokeWithoutArgs;
+namespace {
+
+class MockCaptureListener : public CaptureListener {
+ public:
+  MOCK_METHOD(
+      void, OnCaptureStarted,
+      (int32_t /*process_id*/, std::string /*process_name*/, std::shared_ptr<Process> /*process*/,
+       (absl::flat_hash_map<uint64_t, orbit_client_protos::FunctionInfo>)/*selected_functions*/,
+       TracepointInfoSet /*selected_tracepoints*/),
+      (override));
+  MOCK_METHOD(void, OnCaptureComplete, (), (override));
+  MOCK_METHOD(void, OnCaptureCancelled, (), (override));
+  MOCK_METHOD(void, OnCaptureFailed, (ErrorMessage), (override));
+  MOCK_METHOD(void, OnTimer, (const TimerInfo&), (override));
+  MOCK_METHOD(void, OnKeyAndString, (uint64_t /*key*/, std::string), (override));
+  MOCK_METHOD(void, OnUniqueCallStack, (CallStack), (override));
+  MOCK_METHOD(void, OnCallstackEvent, (CallstackEvent), (override));
+  MOCK_METHOD(void, OnThreadName, (int32_t /*thread_id*/, std::string /*thread_name*/), (override));
+  MOCK_METHOD(void, OnAddressInfo, (LinuxAddressInfo), (override));
+  MOCK_METHOD(void, OnUniqueTracepointInfo, (uint64_t /*key*/, TracepointInfo /*tracepoint_info*/),
+              (override));
+  MOCK_METHOD(void, OnTracepointEvent, (TracepointEventInfo), (override));
+};
+
+TEST(CaptureDeserializer, LoadFileNotExists) {
+  MockCaptureListener listener;
+  std::atomic<bool> cancellation_requested = false;
+  EXPECT_CALL(listener, OnCaptureFailed).Times(1);
+  EXPECT_CALL(listener, OnCaptureStarted).Times(0);
+  EXPECT_CALL(listener, OnCaptureComplete).Times(0);
+  EXPECT_CALL(listener, OnCaptureCancelled).Times(0);
+  capture_deserializer::Load("not_existing_test_file", &listener, &cancellation_requested);
+}
+
+TEST(CaptureDeserializer, LoadNoVersion) {
+  MockCaptureListener listener;
+  std::atomic<bool> cancellation_requested = false;
+  ErrorMessage actual_error;
+  EXPECT_CALL(listener, OnCaptureFailed).Times(1);
```

Comment:
do you want to check parameter here as well? So that you get the correct error message?

Reply:
I was a bit unsure of this. I think it is a bit contradictory to what the test should ensure. The intended behaviour is to have the listener be notified that something went wrong. The exact error message may be well subject to change and should be not that much of the focus of the test.
This is a bit different for the test with the wrong version, because there I expect having an error message that tells me something about the version.
It will also introduce unnecessary work when changing the error message.

But I am not having a very strong opinion on this. So I would be also fine, with adding the check there.

Reply:
I think certain error messages are expected expected in certain situation and it is ok to check them. If they change - they change, as for other things - we fix test as well to match expected behavior.

In this test you are emulating a specific thing that goes wrong. so it is fine to expect a specific error message in this case.

Reply:
Okay, I added the checks.